src/docfx/build/dependency/DependencyMapBuilder.cs

```diff
@@ -34,13 +35,63 @@ public void AddDependencyItem(Document from, Document to, DependencyType type)
 
         public DependencyMap Build()
         {
-            return new DependencyMap(_dependencyItems
-                .GroupBy(k => k.Source)
+            return new DependencyMap(Flatten());
+        }
+
+        private Dictionary<Document, HashSet<DependencyItem>> Flatten()
+        {
+            var result = new Dictionary<Document, HashSet<DependencyItem>>();
+            var graph = _dependencyItems
+                .GroupBy(k => k.From)
                 .ToDictionary(
                     k => k.Key,
-                    v => (from r in v.Distinct()
-                         orderby r.Dest.FilePath, r.Type
-                         select r).ToList()));
+                    v => (from r in v
+                          orderby r.To.FilePath, r.Type
+                          select r).ToHashSet());
+
+            foreach (var (from, value) in graph)
+            {
+                result[from] = new HashSet<DependencyItem>();
+                foreach (var item in value)
+                {
+                    result[from].Add(item);
+                    Transit(from, graph, item, item, result);
+                }
+            }
+            return result;
+        }
+
+        private void Transit(Document from, Dictionary<Document, HashSet<DependencyItem>> graph, DependencyItem current, DependencyItem next, Dictionary<Document, HashSet<DependencyItem>> dependencies)
+        {
+            if (!CanTransit(current))
+            {
+                return;
+            }
+
+            // if the dependency destination is already in the result set, we can reuse it
+            if (next.To != from && dependencies.ContainsKey(next.To))
```

Comment:
Use `TryGetValue` to save another lookup
```suggestion
            if (next.To != from && dependencies.TryGetValue(next.To, out var xxx))
```